### PR TITLE
FloatingPoint: Clean up and document header, format.

### PR DESCRIPTION
**Pull request by aniemetz**
This cleans up the src/util/floatingpoint.h.in header to conform to the
code style guidelines of CVC4. This further attempts to fully document
the header. The main changes (except for added documentation) are
renames of member variables, getting rid of the redundant functions
`FloatingPointSize::exponent()` and `FloatingPointSize::significand()`,
and a more explicit naming of CVC4 wrapper types vs symFPU trait
types. Else, it's mainly reordering and formatting.

### CODE REVIEW

File: src/util/floatingpoint.h.in [hunk 1]
```diff
@@ -20,587 +20,852 @@
...
-    }
-  }; /* struct FloatingPointSizeHashFunction */
+// Inline these!
+inline bool CVC4_PUBLIC validExponentSize(unsigned e) { return e >= 2; }
+inline bool CVC4_PUBLIC validSignificandSize(unsigned s) { return s >= 2; }
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * Floating point sorts are parameterised by two non-zero constants
```
Comment (martin-cs) [inline src/util/floatingpoint.h.in:42]: Not just non-zero but 2 or more.

File: src/util/floatingpoint.h.in [hunk 2]
```diff
@@ -20,587 +20,852 @@
...
+ * giving the width (in bits) of the exponent and significand
+ * (including the hidden bit).
```
Comment (martin-cs) [inline src/util/floatingpoint.h.in:44]: I probably should have added something like "so IEEE-754 single precision, a.k.a. float is described as 8 24".

File: src/util/floatingpoint.h.in [hunk 3]
```diff
@@ -20,587 +20,852 @@
...
+                bool signedBV);
+  FloatingPoint(const FloatingPointSize& size,
+                const RoundingMode& rm,
+                const Rational& r);
 
   /**
-   * As different conversions are different parameterised kinds, there
-   * is a need for different (C++) types for each one.
+   * Create a FP NaN value of given size.
+   * t: The FP size (format).
```
Comment (martin-cs) [inline src/util/floatingpoint.h.in:434]: Why is this `t:`?

File: src/util/floatingpoint.h.in [hunk 4]
```diff
@@ -20,587 +20,852 @@
...
    */
+  static FloatingPoint makeNaN(const FloatingPointSize& size);
+  /**
+   * Create a FP infinity value of given size.
+   * t:    The FP size (format).
+   * sign: True for +oo, false otherwise.
```
Comment (martin-cs) [inline src/util/floatingpoint.h.in:440]: I think `sign` goes straight to the sign bit, so true is negative.

File: src/util/floatingpoint.h.in [hunk 5]
```diff
@@ -20,587 +20,852 @@
...
+  {
+    UnsignedHashFunction< ::CVC4::BitVectorSize> f;
+    return (key ^ 0x46504256) ^ f(fptbv.d_bv_size);
+  }
+}; /* struct FloatingPointToBVHashFunction */
+
+// It is not possible to pack a number without a size to pack to,
+// thus it is difficult to implement this in a sensible way.  Use
+// FloatingPoint instead...
+#if 0
```
Comment (martin-cs) [inline src/util/floatingpoint.h.in:851]: Maybe we should do something better here than `#if 0`

**Review by martin-cs — APPROVED** [review]
I think all the comments are about things I should have done better so ... I wouldn't block this PR on any of them.  If you have time / inclination to fix / improve them it would be great!

File: src/util/floatingpoint.h.in (hunk 5 above)
Comment (aniemetz) [reply src/util/floatingpoint.h.in:851]: Can I remove these?

Comment (martin-cs) [reply src/util/floatingpoint.h.in:851]: I'm fine with them being removed and maybe just a comment saying to use `FloatingPoint`'s output instead.

File: src/util/floatingpoint.h.in (hunk 3 above)
Comment (aniemetz) [reply src/util/floatingpoint.h.in:434]: Good catch!

File: src/util/floatingpoint.h.in (hunk 1 above)
Comment (aniemetz) [reply src/util/floatingpoint.h.in:42]: When we create a FP sort, it's only parameterized over the exponent and significand width:
https://github.com/CVC4/CVC4/blob/master/src/api/cvc4cpp.cpp#L3505

Comment (aniemetz) [reply src/util/floatingpoint.h.in]: Ah wait, now I get what you mean. They must be `>= 2`.

**Pull request merged**